packages/control/lib/runtime-api-client.js

```diff
@@ -0,0 +1,253 @@
+'use strict'
+
+const { tmpdir, platform, EOL } = require('node:os')
+const { join } = require('node:path')
+const { exec, spawn } = require('node:child_process')
+const { readdir } = require('node:fs/promises')
+const { Readable } = require('node:stream')
+const { Client } = require('undici')
+const WebSocket = require('ws')
+const errors = require('./errors.js')
+
+const PLATFORMATIC_TMP_DIR = join(tmpdir(), 'platformatic', 'pids')
+const PLATFORMATIC_PIPE_PREFIX = '\\\\.\\pipe\\platformatic-'
+
+class RuntimeApiClient {
+  #undiciClients = new Map()
+  #webSockets = new Set()
+
+  async getMatchingRuntime (opts) {
+    const runtimes = await this.getRuntimes()
+
+    let runtime = null
+    if (opts.pid) {
+      runtime = runtimes.find(runtime => runtime.pid === parseInt(opts.pid))
+    } else if (opts.name) {
+      runtime = runtimes.find(runtime => runtime.packageName === opts.name)
+    } else if (runtimes.length === 1) {
+      runtime = runtimes[0]
+    }
+    if (!runtime) {
+      throw errors.RuntimeNotFound()
+    }
+    return runtime
+  }
+
+  async getRuntimes () {
+    const runtimePIDs = platform() === 'win32'
+      ? await this.#getWindowsRuntimePIDs()
+      : await this.#getUnixRuntimePIDs()
+
+    const runtimes = []
+    for (const runtimePID of runtimePIDs) {
+      try {
+        const runtimeMetadata = await this.getRuntimeMetadata(runtimePID)
+        runtimes.push(runtimeMetadata)
+      } catch (err) {
+        continue
+      }
+    }
+    return runtimes
+  }
+
+  async getRuntimeMetadata (pid) {
+    const client = this.#getUndiciClient(pid)
+
+    const { statusCode, body } = await client.request({
+      path: '/api/metadata',
+      method: 'GET'
+    })
+
+    if (statusCode !== 200) {
+      const error = await body.text()
+      throw new errors.FailedToGetRuntimeMetadata(error)
+    }
+
+    const metadata = await body.json()
+    return metadata
+  }
+
+  async getRuntimeServices (pid) {
+    const client = this.#getUndiciClient(pid)
+
+    const { statusCode, body } = await client.request({
+      path: '/api/services',
+      method: 'GET'
+    })
+
+    if (statusCode !== 200) {
+      const error = await body.text()
+      throw new errors.FailedToGetRuntimeServices(error)
+    }
+
+    const runtimeServices = await body.json()
+    return runtimeServices
+  }
+
+  async getRuntimeEnv (pid) {
+    const client = this.#getUndiciClient(pid)
+
+    const { statusCode, body } = await client.request({
+      path: '/api/env',
+      method: 'GET'
+    })
+
+    if (statusCode !== 200) {
+      const error = await body.text()
+      throw new errors.FailedToGetRuntimeEnv(error)
+    }
+
+    const runtimeEnv = await body.json()
+    return runtimeEnv
+  }
+
+  async restartRuntime (pid, options = {}) {
```

Comment:
@mcollina WDYT about restarting runtime in this way?

Reply:
How would that work for `control`? Will the new process be spawned as a "child" of the control cli?

Reply:
Exactly

Reply:
sgtm